src/neo/Ledger/Blockchain.cs

```diff
@@ -237,9 +273,32 @@ private VerifyResult OnNewBlock(Block block)
                 // We can store the new block in block_cache and tell the new height to other nodes after Persist().
                 system.LocalNode.Tell(Message.Create(MessageCommand.Ping, PingPayload.Create(Singleton.Height)));
             }
+            else
+            {
+                if (block.Index + 99 >= HeaderHeight)
+                    system.LocalNode.Tell(new LocalNode.RelayDirectly { Inventory = block });
+                if (block.Index == HeaderHeight + 1)
+                    header_cache.Enqueue(block.Header);
+            }
             return VerifyResult.Succeed;
         }
 
+        private void OnNewHeaders(Header[] headers)
+        {
+            if (header_cache.Count >= 10000) return;
```

Comment:
I think we should move this check to `TaskManager`.

Reply:
But if a neighbor sends a header array longer than the remaining space in header cache, it will give rise to problems. I think a check is needed here.

Reply:
Yes. And `TaskManager` should also check it before requesting headers.

Reply:
Done